docs/ACP/ACP.md

```diff
@@ -0,0 +1,545 @@
+- **Metric Store**: Available at `agent://metrics/{agentId}`
+
+Clients can access these using standard resource methods:
+
+```json
+{
+  "id": "resource-req",
+  "jsonrpc": "2.0",
+  "method": "resources/read",
+  "params": {
+    "uri": "agent://directory"
+  }
+}
+```
+
+### 4.2 Tools
+
+ACP defines standard tools for agent operations:
+
+- **agentSearch**: Find agents matching specific criteria
+- **taskAssign**: Assign a task to the best-matching agent
+- **agentEvaluate**: Provide feedback on agent performance
+
+Tool definitions follow standard schema:
+
+```json
+{
+  "name": "agentSearch",
+  "description": "Find agents matching specific criteria",
+  "inputSchema": {
+    "type": "object",
+    "properties": {
+      "capabilities": {
+        "type": "array",
+        "items": { "type": "string" }
+      },
+      "minScore": { "type": "number" },
+      "maxCost": { "type": "number" }
+    }
+  }
+}
+```
+
+### 4.3 LLM Interaction
+
+ACP supports agent-to-agent LLM interactions through a standardized sampling mechanism:
+
+NLIP connection: ACP adopts NLIP's redirect and redirect_response control messages for managing LLM interactions in a federated environment. This allows ACP to leverage NLIP's mechanisms for API key management, LLM service discovery, and response aggregation. Both protocols can benefit from a unified approach to LLM interaction with proper attribution and security.
+
+```json
+{
+  "id": "sample-req",
+  "jsonrpc": "2.0",
+  "method": "sampling/createMessage",
+  "params": {
+    "messages": [
+      {
+        "role": "user",
+        "content": {
+          "type": "text",
+          "text": "Analyze the performance metrics for agent-id-1"
+        }
+      }
+    ],
+    "maxTokens": 1000,
+    "temperature": 0.7,
+    "modelPreferences": { // TBD
+      "intelligencePriority": 0.8,
+      "speedPriority": 0.6,
+      "costPriority": 0.4
+    }
+  }
+}
+```
+
+## 5. Platform Services
+
+ACP defines several platform services that coordinate multi-agent systems:
+
+### 5.1 Agent Directory Service
+
+- Maintains a registry of active agents
+- Exposes resources and tools for agent discovery
+- Computes agent rankings based on evaluation data
+
+### 5.2 Coordination Service
+
+- Manages task distribution based on capability matching
+- Implements load balancing using agent metrics
+- Provides tools for task management
+
+### 5.3 Security Service
+
+- Validates agent authentication tokens
+- Enforces access control
+- Ensures user consent is maintained for all operations
+
+### 5.4 Monitoring Service
+
+- Tracks agent availability via heartbeats
+- Collects performance metrics
+- Provides resources for system status
+
+## 6. Backward Compatibility
+
+### 6.1 Model Context Protocol (MCP) Compatibility
+
+ACP is designed as a backward-compatible extension to the Model Context Protocol (MCP). This ensures systems can adopt ACP incrementally while maintaining compatibility with existing implementations.
```

Comment:
perhaps add an emphasizing note such as "This means both MCP and ACP messages share the same transport, such as stdio" https://modelcontextprotocol.io/specification/2025-03-26/basic/transports

If that's not intended, it would be great to highlight, as I thought about this when reading